tools/autograd/templates/Functions.cpp

```diff
@@ -1800,44 +1800,127 @@ Tensor qr_backward(const std::vector<torch::autograd::Variable> &grads, const Te
-    auto gsigma = grad.div(sigma);
+    // logdet = \sum log(sigma)
+    auto gsigma = grad.unsqueeze(-1).div(sigma);
     return svd_backward({{}, gsigma, {}}, self, true, true, u, sigma, v);
+  };
+
+  auto nonsingular_case_backward = [&](const Tensor& grad, const Tensor& self) -> Tensor {
+    return unsqueeze_multiple(grad, {-1, -2}, self.dim()) * self.inverse().transpose(-2, -1);
+  };
+
+  if (self.dim() == 2) {
+    if ((logdet != -INFINITY).item<uint8_t>()) {
+      return nonsingular_case_backward(grad, self);
+    } else {
+      return singular_case_backward(grad, self);
+    }
+  } else {
+    Tensor grad_logdet = at::empty_like(self);
+    auto finite_logdet_indices = at::where(logdet != -INFINITY);
+    auto neginf_logdet_indices = at::where(logdet == -INFINITY);
+
+    if (finite_logdet_indices[0].size(0) > 0) {  // invertible case
+      auto indexed_grad = grad.index(finite_logdet_indices);
+      auto indexed_self = self.index(finite_logdet_indices);
+      grad_logdet.index_put_(/*indices=*/finite_logdet_indices,
+                          /*value=*/nonsingular_case_backward(indexed_grad, indexed_self));
+    }
+    if (neginf_logdet_indices[0].size(0) > 0) {  // non-invertible case, uses SVD
+      auto indexed_grad = grad.index(neginf_logdet_indices);
+      auto indexed_self = self.index(neginf_logdet_indices);
+      grad_logdet.index_put_(/*indices=*/neginf_logdet_indices,
+                          /*value=*/singular_case_backward(indexed_grad, indexed_self));
+    }
+    return grad_logdet;
   }
 }
 
 Tensor slogdet_backward(const Tensor& grad_logabsdet,
                         const Tensor& self,
                         const Tensor& signdet, const Tensor& logabsdet) {
-  auto signdet_val = signdet.item<double>();
-  if (signdet_val != 0 /* det != 0, invertible */) {
-    return grad_logabsdet * self.inverse().t();
-  } else /* otherwise det = \prod(sigma) = 0, use svd */ {
+  auto singular_case_backward = [&](const Tensor& grad_logabsdet, const Tensor& self) -> Tensor {
     Tensor u, sigma, v;
     std::tie(u, sigma, v) = self.svd();
     // sigma has all non-negative entries (also with at least one zero entry)
     // so logabsdet = \sum log(abs(sigma))
     // but det = 0, so backward logabsdet = \sum log(sigma)
-    auto gsigma = grad_logabsdet.div(sigma);
+    auto gsigma = grad_logabsdet.unsqueeze(-1).div(sigma);
     return svd_backward({{}, gsigma, {}}, self, true, true, u, sigma, v);
+  };
+
+  auto nonsingular_case_backward = [&](const Tensor& grad_logabsdet, const Tensor& self) -> Tensor {
+    return unsqueeze_multiple(grad_logabsdet, {-1, -2}, self.dim()) * self.inverse().transpose(-2, -1);
+  };
+
+  if (self.dim() == 2) {
+    if ((signdet == 0).item<uint8_t>()) {
+      return singular_case_backward(grad_logabsdet, self);
+    } else {
+      return nonsingular_case_backward(grad_logabsdet, self);
+    }
+  } else {
+    Tensor grad_slogdet = at::empty_like(self);
```

Comment:
Similarly, can you optimize also for the case where all dets are zero or nonzero and an extra `index_put` copy is not needed?